File: tests/layer0_ops/unit/test_math_utils.py
"""
test_math_utils.py
==================
Python port of:
    layer0_ops/unit/TestMathUtils.m

Covers: pderiv, psech

Test categories
---------------
  [Reference]  Always runs – uses numpy reference implementations from conftest.
  [Impl]       @pytest.mark.requires_impl – calls cool_frames.layer0.
"""

from __future__ import annotations

import pytest

import numpy as np
from conftest import pderiv_ref, psech_ref

# ---------------------------------------------------------------------------
# pderiv – reference tests (always run)
# ---------------------------------------------------------------------------

class TestPderivReference:
    """
    Structural and mathematical tests of pderiv_ref().

    MATLAB counterpart: TestMathUtils (pderiv section).
    """

    def test_output_length_vector(self):
        """pderiv must return a vector of the same length as input."""
        x  = np.random.default_rng(0).standard_normal(64)
        fd = pderiv_ref(x)
        assert len(fd) == len(x)

    def test_output_size_matrix(self):
        """For a 2-D input, pderiv_ref acts along axis-0 (columns)."""
        X  = np.random.default_rng(1).standard_normal((64, 4))
        # Apply column-wise
        FD = np.column_stack([pderiv_ref(X[:, j]) for j in range(X.shape[1])])
        assert FD.shape == X.shape

    def test_derivative_of_constant_is_zero(self):
        """Derivative of a constant signal is zero for all diff orders."""
        L = 64
        f = np.ones(L)
        for order in (2, 4, np.inf):
            fd = pderiv_ref(f, difforder=order)
            assert np.linalg.norm(fd) < 1e-10, (
                f"pderiv_ref(const, order={order}): norm = {np.linalg.norm(fd):.2e} > 1e-10"
            )

    def test_derivative_of_sine(self):
        """
        Derivative of sin(2π·k·x) on [0,1) = 2π·k · cos(2π·k·x).

        MATLAB counterpart: testPderivLinearSine
        """
        L = 128
        n = np.arange(L)
        f        = np.sin(2 * np.pi * n / L)
        fd       = pderiv_ref(f, difforder=np.inf)
        expected = 2 * np.pi * np.cos(2 * np.pi * n / L)
        rel_err  = np.linalg.norm(fd - expected) / np.linalg.norm(expected)
        assert rel_err < 1e-10, (
            f"pderiv_ref(sin, Inf): rel_err={rel_err:.2e} > 1e-10"
        )

    def test_order_inf_real_for_real_input(self):
        """Spectral derivative of a real signal must be real."""
        f  = np.random.default_rng(2).standard_normal(64)
        fd = pderiv_ref(f, difforder=np.inf)
        assert np.isrealobj(fd), "pderiv_ref(Inf): output must be real for real input"

    def test_order2_vs_order4_smooth_signal(self):
        """For a low-frequency signal, orders 2 and 4 agree within 2 %."""
        L  = 256
        n  = np.arange(L)
        f  = np.sin(2 * np.pi * 3 * n / L) + 0.5 * np.cos(2 * np.pi * 5 * n / L)
        fd2 = pderiv_ref(f, difforder=2)
        fd4 = pderiv_ref(f, difforder=4)
        rel_err = np.linalg.norm(fd2 - fd4) / (np.linalg.norm(fd4) + np.finfo(float).eps)
        assert rel_err < 0.02, (
            f"pderiv_ref orders 2 vs 4: rel_err={rel_err:.3f} > 0.02"
        )

    @pytest.mark.parametrize("order", [2, 4])
    def test_real_output_for_real_input(self, order):
        """Finite-difference orders must return real for real input."""
        f  = np.random.default_rng(3).standard_normal(64)
        fd = pderiv_ref(f, difforder=order)
        assert np.isrealobj(fd), f"pderiv_ref(order={order}): real input must give real output"


# ---------------------------------------------------------------------------
# psech – reference tests (always run)
# ---------------------------------------------------------------------------

class TestPsechReference:
    """
    Structural and mathematical tests of psech_ref().

    MATLAB counterpart: TestMathUtils (psech section).
    """

    @pytest.mark.parametrize("L", [32, 64, 128, 256])
    def test_output_length(self, L):
        """psech must return a vector of length L."""
        g = psech_ref(L)
        assert len(g) == L

    @pytest.mark.parametrize("L", [32, 64, 128])
    def test_unit_norm(self, L):
        """psech is normalised to unit L2-norm."""
        g = psech_ref(L)
        assert abs(np.linalg.norm(g) - 1.0) < 1e-4, (
            f"psech_ref(L={L}): norm={np.linalg.norm(g):.6f} ≠ 1.0"
        )

    def test_real_valued(self):
        """psech must return a real-valued signal."""
        g = psech_ref(64)
        assert np.isrealobj(g)

    def test_dft_invariance(self):
        """
        The canonical psech(L, tfr=1) is its own DFT:
            norm(g - fft(g)/sqrt(L)) ≈ 0

        MATLAB counterpart: testPsechDFTInvariance (uses LTFAT dft = fft/sqrt(L))
        """
        L   = 128
        g   = psech_ref(L)
        dft = np.fft.fft(g) / np.sqrt(L)    # LTFAT dft convention
        err = np.linalg.norm(g - dft)
        assert err < 1e-8, f"psech_ref DFT invariance: err={err:.2e} > 1e-8"

    @pytest.mark.parametrize("L", [32, 64, 128])
    def test_fft_real_valued(self, L):
        """fft(psech) must be real (whole-point even window)."""
        g = psech_ref(L)
        G = np.fft.fft(g)
        assert np.max(np.abs(np.imag(G))) < 1e-12, (
            f"psech_ref(L={L}): fft not real (max imag = {np.max(np.abs(np.imag(G))):.2e})"
        )

    def test_tfr_scaling(self):
        """
        Larger tfr → wider time window → lower peak value.

        MATLAB counterpart: testPsechTfrScaling
        """
        L  = 128
        g1 = psech_ref(L, tfr=1)
        g2 = psech_ref(L, tfr=4)
        assert np.max(np.abs(g2)) < np.max(np.abs(g1)), (
            "psech_ref: tfr=4 peak must be lower than tfr=1 peak"
        )

File: tests/layer0_ops/unit/TestMathUtils.m
classdef TestMathUtils < matlab.unittest.TestCase
%TESTMATHUTILS  Unit tests for mathematical utility functions.
%
%   Covers: pderiv, psech
%
%   pderiv(f)            -- periodic numerical/spectral derivative
%   psech(L)             -- periodized hyperbolic-secant window

    methods (TestClassSetup)
        function setupClass(tc)
            sharedDir = fullfile(fileparts(mfilename('fullpath')), ...
                                 '..', '..', 'shared');
            addpath(sharedDir);
            setup_filterbank_paths();
        end
    end

    % ── pderiv: output dimensions ─────────────────────────────────────────
    methods (Test)

        function testPderivOutputLengthVector(tc)
            % pderiv must return a vector of the same length as input.
            x = randn(64, 1);
            fd = pderiv(x);
            tc.verifyEqual(numel(fd), numel(x), ...
                'pderiv: output length must equal input length.');
        end

        function testPderivOutputSizeMatrix(tc)
            % When applied to a matrix, pderiv acts along columns by default.
            X = randn(64, 4);
            FD = pderiv(X);
            tc.verifyEqual(size(FD), size(X), ...
                'pderiv: output size must equal input size for a matrix.');
        end

        function testPderivDerivativeOfConstantIsZero(tc)
            % Derivative of a constant signal is zero (all orders).
            L = 64;
            f = ones(L, 1);
            for order = [2, 4, Inf]
                fd = pderiv(f, [], order);
                tc.verifyLessThan(norm(fd), 1e-10, ...
                    sprintf('pderiv: derivative of constant must be zero (order %g).', order));
            end
        end

        function testPderivLinearSine(tc)
            % Derivative of sin(2*pi*k*x) = 2*pi*k * cos(2*pi*k*x).
            % On [0,1) with L samples: f(n) = sin(2*pi*n/L),
            % f'(n) = (2*pi/L) * cos(2*pi*n/L) scaled by L because pderiv
            % returns the derivative on [0,1) (not [0,L)).
            % pderiv returns d/d(n/L), so derivative = 2*pi * cos(2*pi*n/L).
            L = 128;
            n = (0:L-1).';
            f  = sin(2*pi*n/L);
            fd = pderiv(f, [], Inf);          % spectral derivative
            expected = 2*pi * cos(2*pi*n/L);
            rel_err = norm(fd - expected) / norm(expected);
            tc.verifyLessThan(rel_err, 1e-10, ...
                'pderiv(Inf): derivative of sin must equal 2*pi*cos (spectral).');
        end

        function testPderivOrderInf(tc)
            % Spectral derivative (order=Inf) of a real signal is real.
            f = randn(64, 1);
            fd = pderiv(f, [], Inf);
            tc.verifyTrue(isreal(fd), ...
                'pderiv(Inf): derivative of a real signal must be real.');
        end

        function testPderivOrder2VsOrder4Agreement(tc)
            % For a smooth low-frequency signal, orders 2 and 4 give close results.
            L = 256;
            n = (0:L-1).';
            f = sin(2*pi*3*n/L) + 0.5*cos(2*pi*5*n/L);
            fd2 = pderiv(f, [], 2);
            fd4 = pderiv(f, [], 4);
            rel_err = norm(fd2 - fd4) / (norm(fd4) + eps);
            tc.verifyLessThan(rel_err, 0.02, ...
                'pderiv: orders 2 and 4 must agree within 2% for smooth signals.');
        end

        function testPderivRealOutputForRealInput(tc)
            % All finite difference orders must return real output for real input.
            f = randn(64, 1);
            for order = [2, 4]
                fd = pderiv(f, [], order);
                tc.verifyTrue(isreal(fd), ...
                    sprintf('pderiv(order=%d): real input must give real output.', order));
            end
        end

    end

    % ── psech: output dimensions and basic properties ─────────────────────
    methods (Test)

        function testPsechOutputLength(tc)
            % psech(L) must return a column vector of length L.
            for L = [32, 64, 128, 256]
                g = psech(L);
                tc.verifyEqual(numel(g), L, ...
                    sprintf('psech: output length must equal L=%d.', L));
                tc.verifyEqual(size(g, 2), 1, ...
                    sprintf('psech: output must be a column vector (L=%d).', L));
            end
        end

        function testPsechUnitNorm(tc)
            % psech is normalized so that norm(g) = 1.
            for L = [32, 64, 128]
                g = psech(L);
                tc.verifyEqual(norm(g), 1, 'AbsTol', 1e-4, ...
                    sprintf('psech: norm must equal 1 for L=%d.', L));
            end
        end

        function testPsechIsRealValued(tc)
            % psech must return a real-valued signal.
            g = psech(64);
            tc.verifyTrue(isreal(g), 'psech: output must be real-valued.');
        end

        function testPsechDFTInvariance(tc)
            % The canonical psech with tfr=1 is its own DFT:
            % norm(g - dft(g)) should be close to zero.
            L = 128;
            g = psech(L);
            err = norm(g - dft(g));
            tc.verifyLessThan(err, 1e-10, ...
                'psech: canonical psech(L) must be its own DFT (DFT-invariant).');
        end

        function testPsechWholePointEven(tc)
            % psech is whole-point even: fft(psech(L)) is real.
            for L = [32, 64, 128]
                g = psech(L);
                G = fft(g);
                tc.verifyLessThan(max(abs(imag(G))), 1e-12, ...
                    sprintf('psech: fft must be real-valued (whole-point even), L=%d.', L));
            end
        end

        function testPsechTfrScaling(tc)
            % With tfr > 1, the window should be wider (higher concentration in time).
            L = 128;
            g1 = psech(L, 1);
            g2 = psech(L, 4);
            % Wider window in time → lower peak value
            tc.verifyLessThan(max(abs(g2)), max(abs(g1)), ...
                'psech: larger tfr must give a wider (higher-peak) window.');
        end

    end

end
